File: backend/src/main/java/com/offworld/model/Ship.java
package com.offworld.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import java.util.Map;

// Un vaisseau de transport (trucking ou trade)
@JsonIgnoreProperties(ignoreUnknown = true)
public record Ship(
        String id,
        String ownerId,
        String originPlanetId,
        String destinationPlanetId,
        Map<String, Long> cargo,
        String status,
        String tradeId,
        String truckingId,
        Long fee,
        long createdAt,
        Long arrivalAt,
        Long operationCompleteAt
) {
    // Les statuts possibles côté serveur (snake_case)
    public static final String IN_TRANSIT_TO_ORIGIN        = "in_transit_to_origin";
    public static final String AWAITING_ORIGIN_DOCKING_AUTH = "awaiting_origin_docking_auth";
    public static final String LOADING                     = "loading";
    public static final String AWAITING_ORIGIN_UNDOCKING_AUTH = "awaiting_origin_undocking_auth";
    public static final String IN_TRANSIT                  = "in_transit";
    public static final String AWAITING_DOCKING_AUTH       = "awaiting_docking_auth";
    public static final String UNLOADING                   = "unloading";
    public static final String AWAITING_UNDOCKING_AUTH     = "awaiting_undocking_auth";
    public static final String COMPLETE                    = "complete";

    public boolean needsOriginDock() {
        return AWAITING_ORIGIN_DOCKING_AUTH.equals(status);
    }

    public boolean needsOriginUndock() {
        return AWAITING_ORIGIN_UNDOCKING_AUTH.equals(status);
    }

    public boolean needsDestDock() {
        return AWAITING_DOCKING_AUTH.equals(status);
    }

    public boolean needsDestUndock() {
        return AWAITING_UNDOCKING_AUTH.equals(status);
    }

    public boolean isDone() {
        return COMPLETE.equals(status);
    }
}

File: server/src/models/ship.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum ShipStatus {
    InTransitToOrigin,
    AwaitingOriginDockingAuth,
    Loading,
    AwaitingOriginUndockingAuth,
    InTransit,
    AwaitingDockingAuth,
    Unloading,
    AwaitingUndockingAuth,
    Complete,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct Ship {
    pub id: Uuid,
    pub owner_id: String,
    pub origin_planet_id: String,
    pub destination_planet_id: String,
    pub cargo: HashMap<String, u64>,
    pub status: ShipStatus,
    pub trade_id: Option<Uuid>,
    pub trucking_id: Option<Uuid>,
    pub fee: Option<u64>,
    pub created_at: u64,
    pub arrival_at: Option<u64>,
    pub operation_complete_at: Option<u64>,
    #[serde(default)]
    pub estimated_arrival_at: Option<u64>,
    #[serde(default)]
    pub callback_url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema, Validate)]
pub struct CreateTruckingRequest {
    #[validate(length(min = 1, max = 128))]
    pub destination_planet_id: String,
    #[validate(length(min = 1, max = 128))]
    pub origin_planet_id: String,
    pub cargo: HashMap<String, u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct DockRequest {
    pub authorized: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct UndockRequest {
    pub authorized: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ShipWebhookPayload {
    OriginDockingRequest {
        ship_id: Uuid,
        origin_planet_id: String,
        destination_planet_id: String,
        cargo: HashMap<String, u64>,
    },
    DockingRequest {
        ship_id: Uuid,
        origin_planet_id: String,
        cargo: HashMap<String, u64>,
    },
    ShipDocked {
        ship_id: Uuid,
        status: String,
    },
    ShipComplete {
        ship_id: Uuid,
    },
}
